pruner/pruner.go

```diff
@@ -0,0 +1,339 @@
+package pruner
+
+import (
+	"sync"
+	"time"
+
+	quamina "quamina/lib"
+)
+
+// Stats reports basic counts to aid in deciding when to Rebuild.
+type Stats struct {
+	// Some of these values are ints instead of uints because Go
+	// likes to print uints in hex, and I'd like to see some
+	// temporary logging output in decimal.  ToDo: back to uints?
+
+	// Live the count of total live patterns.
+	Live int
+
+	// Added is the count of total patterns added.
+	Added int
+
+	// Deleted is the count of the patterns removed.
+	Deleted int
+
+	// Emitted is the count to total patterns found since the last
+	// rebuild.
+	Emitted int64
+
+	// Filtered is the count of pattners that have been removed
+	// from MatchFor results (since the last rebuild) because
+	// their patterns had been removed.
+	Filtered int64
+
+	// LastRebuilt is the time the last rebuild started.
+	LastRebuilt time.Time
+
+	// RebuildDuration is the duration of the last rebuild.
+	RebuildDuration time.Duration
+
+	// RebuildPurged is the count of patterns removed during
+	// rebuild.
+	RebuildPurged int
+}
+
+// Matcher provides DelPattern on top of quamina.Matcher.
+//
+// Matcher maintains the set of live patterns, and it will rebuild the
+// underlying matcher synchronously (currently) periodically sduring
+// standard operations (AddPattern, DelPattern, MatchesForJSONEvent).
+//
+// Roughly speaking, the current rebuild policy automatically rebuilds
+// the index when the ratio of filtered patterns to emitted patterns
+// exceeds 0.2 (and if there's been some traffic).
+//
+// An application can call Rebuild to force a rebuild at any any.  See
+// Stats() to obtain some useful statistics about the matcher.
+//
+// Eventually automatically-invoked rebuild policies might be
+// pluggable.
+type Matcher struct {
+	// Matcher is the underlying matcher that does the hard work.
+	//
+	// Maybe Matcher should maybe not be embedded or public.
+	*quamina.Matcher
+
+	// live is live set of patterns.
+	live State
+
+	stats Stats
+
+	// rebuildTrigger, if not nil, determines when a mutation
+	// triggers a rebuild.
+	//
+	// If nil, no automatic Rebuild is ever triggered.
+	rebuildTrigger rebuildTrigger
+
+	// lock protectes the pointer the underlying Matcher as well as stats.
+	//
+	// The Matcher pointer is updated after a successful Rebuild.
+	// Stats are updated by Add, Del, and Rebuild.
+	lock sync.RWMutex
+}
+
+var defaultRebuildTrigger = newTooMuchFiltering(0.2, 1000)
+
+// tooMuchFiltering is the standard rebuildTrigger, which will fire
+// when:
+//
+//   MinAction is less than the sum of counts of found and filtered
+//   patterns and
+//
+//   FilteredToEmitted is greater than the ratio of counts of filtered
+//   and found patterns.
+//
+// defaultRebuildTrigger provides the default trigger policy used by
+// NewMatcher.
+type tooMuchFiltering struct {
```

Comment:
why *64?  I like to constrain the compiler as little as possible unless the semantics get in the way… I'd probably just say `float` and `int`.  Seems unlikely to me that anyone would want to require more than 2**32 matches…

Reply:
After thinking about the type of the `Emitted` counter based on your comments, seems like we might in fact want `int64` here so that there is no cast in the arithmetic. Since an `Emitted int` could reasonably overflow on some platforms, seems like `Emitted` should be an `int64`. Keep these types consistent. (I'll resist arguing for an `uint64` I hope.)

Reply:
sure